File: src/main/java/svnserver/repository/git/GitFileEmptyTree.java
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.repository.git;

import org.eclipse.jgit.lib.FileMode;
import org.eclipse.jgit.lib.ObjectId;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import svnserver.repository.VcsCopyFrom;
import svnserver.repository.git.filter.GitFilter;
import svnserver.repository.git.prop.GitProperty;

import java.io.InputStream;
import java.util.Collections;

/**
 * Git file.
 *
 * @author Artem V. Navrotskiy <bozaro@users.noreply.github.com>
 */
final class GitFileEmptyTree extends GitEntryImpl implements GitFile {
  @NotNull
  private final GitBranch branch;

  private final int revision;

  GitFileEmptyTree(@NotNull GitBranch branch, @NotNull String parentPath, int revision) {
    super(GitProperty.emptyArray, parentPath, GitProperty.emptyArray, "", FileMode.TREE);
    this.branch = branch;
    this.revision = revision;
  }

  @Nullable
  @Override
  public GitFile getEntry(@NotNull String name) {
    return null;
  }

  @NotNull
  @Override
  public String getContentHash() {
    throw new IllegalStateException("Can't get content hash without object.");
  }

  @NotNull
  @Override
  public String getMd5() {
    throw new IllegalStateException("Can't get md5 without object.");
  }

  @Override
  public long getSize() {
    return 0L;
  }

  @NotNull
  @Override
  public InputStream openStream() {
    throw new IllegalStateException("Can't get open stream without object.");
  }

  @Nullable
  @Override
  public VcsCopyFrom getCopyFrom() {
    return getLastChange().getCopyFrom(getFullPath());
  }

  @Nullable
  public GitFilter getFilter() {
    return null;
  }

  @Nullable
  public GitTreeEntry getTreeEntry() {
    return null;
  }

  @Nullable
  public GitObject<ObjectId> getObjectId() {
    return null;
  }

  @NotNull
  @Override
  public GitBranch getBranch() {
    return branch;
  }

  @Override
  public int getRevision() {
    return revision;
  }

  @NotNull
  public FileMode getFileMode() {
    return FileMode.TREE;
  }

  @NotNull
  @Override
  public Iterable<GitFile> getEntries() {
    return Collections.emptyList();
  }

  @Override
  public String toString() {
    return "GitFileEmptyTree{" +
        "fullPath='" + getFullPath() + '\'' +
        '}';
  }
}

File: src/main/kotlin/svnserver/repository/git/GitFileEmptyTree.kt
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.repository.git

import org.eclipse.jgit.lib.FileMode
import org.eclipse.jgit.lib.ObjectId
import svnserver.repository.VcsCopyFrom
import svnserver.repository.git.filter.GitFilter
import svnserver.repository.git.prop.GitProperty
import java.io.InputStream
import java.util.*
import java.util.function.Supplier

/**
 * Git file.
 *
 * @author Artem V. Navrotskiy <bozaro@users.noreply.github.com>
 */
internal class GitFileEmptyTree(override val branch: GitBranch, parentPath: String, override val revision: Int) : GitEntryImpl(GitProperty.emptyArray, parentPath, GitProperty.emptyArray, "", FileMode.TREE, branch.repository.context.shared.stringInterner), GitFile {
    override val contentHash: String
        get() {
            throw IllegalStateException("Can't get content hash without object.")
        }
    override val md5: String
        get() {
            throw IllegalStateException("Can't get md5 without object.")
        }
    override val size: Long
        get() {
            return 0L
        }

    override fun openStream(): InputStream {
        throw IllegalStateException("Can't get open stream without object.")
    }

    override val copyFrom: VcsCopyFrom?
        get() {
            return lastChange.getCopyFrom(fullPath)
        }
    override val filter: GitFilter?
        get() {
            return null
        }
    override val treeEntry: GitTreeEntry?
        get() {
            return null
        }
    override val objectId: GitObject<ObjectId>?
        get() {
            return null
        }
    override val fileMode: FileMode
        get() {
            return FileMode.TREE
        }

    override val entries: SortedMap<String, Supplier<GitFile>>
        get() {
            return emptyEntries
        }

    override fun toString(): String {
        return "GitFileEmptyTree{fullPath='$fullPath'}"
    }

    private companion object {
        private val emptyEntries = Collections.unmodifiableSortedMap(TreeMap<String, Supplier<GitFile>>())
    }
}
